flytepropeller/pkg/controller/nodes/array/handler.go

```diff
@@ -576,18 +576,20 @@ func (a *arrayNodeHandler) Handle(ctx context.Context, nCtx interfaces.NodeExecu
 			taskPhase = idlcore.TaskExecution_FAILED
 		}
 
-		// if the ArrayNode phase has changed we need to reset the taskPhaseVersion to 0, otherwise
-		// increment it if we detect any changes in subNode state.
-		if currentArrayNodePhase != arrayNodeState.Phase {
-			arrayNodeState.TaskPhaseVersion = 0
-		} else if incrementTaskPhaseVersion {
+		// increment taskPhaseVersion if we detect any changes in subNode state.
+		if incrementTaskPhaseVersion {
 			arrayNodeState.TaskPhaseVersion = arrayNodeState.TaskPhaseVersion + 1
 		}
 
 		if err := eventRecorder.finalize(ctx, nCtx, taskPhase, arrayNodeState.TaskPhaseVersion, a.eventConfig); err != nil {
 			logger.Errorf(ctx, "ArrayNode event recording failed: [%s]", err.Error())
 			return handler.UnknownTransition, err
 		}
+
+		// if the ArrayNode phase has changed we need to reset the taskPhaseVersion to 0
+		if currentArrayNodePhase != arrayNodeState.Phase {
+			arrayNodeState.TaskPhaseVersion = 0
+		}
```

Comment:
Doesn't setting this afterwards cause problems with admin requiring incremental values?

Reply:
The scenario of interest would be if we would lose some eventing data if we have incrementPhase = False and currentArrayNodePhase != arrayNodeState.Phase + have a previous event emitted with the same TaskPhase and TaskPhaseVersion.

We set incrementTaskPhaseVersion = True if there's a subnode phase change. The arrayNodeState.Phase is updated in 3 different places: [arrayNodeState.Phase = v1alpha1.ArrayNodePhaseFailing](https://github.com/flyteorg/flyte/blob/25c35961f7d39b75868c0cad312984d73aa245ae/flytepropeller/pkg/controller/nodes/array/handler.go#L404), [arrayNodeState.Phase = v1alpha1.ArrayNodePhaseSucceeding](https://github.com/flyteorg/flyte/blob/25c35961f7d39b75868c0cad312984d73aa245ae/flytepropeller/pkg/controller/nodes/array/handler.go#L407), and [arrayNodeState.Phase = v1alpha1.ArrayNodePhaseExecuting](https://github.com/flyteorg/flyte/blob/25c35961f7d39b75868c0cad312984d73aa245ae/flytepropeller/pkg/controller/nodes/array/handler.go#L407).

For arrayNodeState.Phase = v1alpha1.ArrayNodePhaseFailing and arrayNodeState.Phase = v1alpha1.ArrayNodePhaseSucceeding, there would have to be subnode phase change so we couldn't have a scenario where incrementPhase = False and currentArrayNodePhase != arrayNodeState.Phase.

For arrayNodeState.Phase = v1alpha1.ArrayNodePhaseExecuting, we shouldn't have a previous event emitted with the same TaskPhase and TaskPhaseVersion as we should only be in the v1alpha1.ArrayNodePhaseNone phase for the first pass through.

Reply:
The proceeding loops would have a new TaskPhase as well.

Reply:
> The scenario of interest would be if we would lose some eventing data if we have incrementPhase = False and currentArrayNodePhase != arrayNodeState.Phase + have a previous event emitted with the same TaskPhase and TaskPhaseVersion.
Do we have a repro for this? I'm having difficulty understanding how this is possible. If `incrementPhase == False` then no subnode phases have been updated so `currentArrayNodePhase != arrayNodeState.Phase` cannot be true since they are determined from subNode phases - unless this is a system failure in the handler logic?

Then `taskPhase` is deterministic on `currentArrayNodePhase`, so if `currentArrayNodePhase != arrayNodeState.Phase` then we can not have emitted an event with the same `taskPhase` and `taskPhaseVersion`. FlyteAdmin does not take `taskPhaseVersion` into accord for new `taskPhase` values.

I'm probably missing something here, a repro would help.

Reply:
@hamersaw I noticed this when working on mapping over launch plans but don't think I actually ran into the bug. However this should repro an issue.

```
@task
def hello(num: int) -> int:
    if num > 9:
        time.sleep(10)
        raise Exception("This is a test exception")
    return num


@workflow
def map_workflow():
    map_task(hello, min_success_ratio=0.5)(num=[1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
```

A subnode task phase will be stuck in "Running" even though it has terminated/failed.

Emitting the event always fails due to the task phase not getting bumped but the task phase version getting reset.

Reply:
Wait, this doesn't occur when the staggered subtask succeeds. Looking back into this.

Reply:
Figured it out:

this doesn't bubble up for:

```
@task
def hello(num: int) -> int:
    if num > 9:
        time.sleep(10)
    return num


@workflow
def map_workflow():
    map_task(hello, min_success_ratio=0.5)(num=[1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
```

https://github.com/flyteorg/flyte/blob/fc1c92cce952f26242b60b59ef4e78d61f37e69c/flytepropeller/pkg/controller/nodes/array/handler.go#L374-L386 doesn't check for NodePhaseSucceeding so it counts as a running task so we don't bump the arrayNode phase here: https://github.com/flyteorg/flyte/blob/fc1c92cce952f26242b60b59ef4e78d61f37e69c/flytepropeller/pkg/controller/nodes/array/handler.go#L403-L409

Since we don't do that we bump task phase version instead of resetting to 0 to the event lands in admin. Meanwhile for NodePhaseFailing, we don't increment running so we update arrayNodeState.Phase  which then leads for the existing implementation to reset the task phase version.